design/hotshard.md

```diff
@@ -0,0 +1,385 @@
+# Hot Shard Management in FoundationDB
+
+## Table of Contents
+
+- [Overview](#overview)
+- [What is a Hot Shard?](#what-is-a-hot-shard)
+- [Normal Internal Shard Splitting Process](#normal-internal-shard-splitting-process)
+- [Hot Shard Special Case](#hot-shard-special-case)
+- [When a Hot Shard Occurs That Cannot Be Split](#when-a-hot-shard-occurs-that-cannot-be-split)
+  - [Storage Server Impact](#storage-server-impact)
+  - [Cluster-Wide Rate Limiting](#cluster-wide-rate-limiting)
+- [Mitigation](#mitigation)
+  - [Avoiding Hot Shards](#avoiding-hot-shards)
+  - [Avoiding Read Hotspots](#avoiding-read-hotspots)
+  - [Schema & Keyspace Design Patterns](#schema--keyspace-design-patterns)
+  - [What to do when encountered in production](#what-to-do-when-encountered-in-production)
+  - [Optional Hot Shard Throttling (Targeted Mitigation)](#optional-hot-shard-throttling-targeted-mitigation)
+  - [Key Insight](#key-insight)
+- [Server Knobs Related to Hot Shards](#server-knobs-related-to-hot-shards)
+  - [Hot Shard Throttling Knobs](#hot-shard-throttling-knobs)
+  - [Shard Bandwidth Splitting Knobs](#shard-bandwidth-splitting-knobs)
+  - [Read Hot Shard Detection Knobs](#read-hot-shard-detection-knobs)
+  - [Storage Queue Rebalancing Knobs](#storage-queue-rebalancing-knobs)
+  - [Read Rebalancing Knobs](#read-rebalancing-knobs)
+  - [Shard Split/Merge Priority Knobs](#shard-splitmerge-priority-knobs)
+  - [Other Hot Shard Related Knobs](#other-hot-shard-related-knobs)
+  - [Key Relationships](#key-relationships)
+- [References](#references)
+
+## Overview
+
+FoundationDB partitions the keyspace into shards (contiguous key ranges) and assigns each shard to a storage-team replica set.
+
+Shards are just key ranges. Their size adapts continuously as Data Distribution monitors load and splits or merges ranges to stay within configured bounds [1][2]. Splits are triggered when metrics like bytes, read/write bandwidth, or hot-shard detection exceed thresholds (splitMetrics.bytes = shardBounds.max.bytes / 2, splitMetrics.bytesWrittenPerKSecond etc.). In other words, shard size is workload driven rather than fixed [1][3]. Storage servers and commit proxies provide ongoing measurements so DD can resize ranges dynamically to relieve load hotspots and maintain balance [4][5].
```

Comment:
```suggestion
Shards are just key ranges. Their size adapts continuously as Data Distributor (DD) monitors load and splits or merges ranges to stay within configured bounds [1][2]. Splits are triggered when metrics like bytes, read/write bandwidth, or hot-shard detection exceed thresholds (splitMetrics.bytes = shardBounds.max.bytes / 2, splitMetrics.bytesWrittenPerKSecond etc.). In other words, shard size is workload driven rather than fixed [1][3]. Storage servers and commit proxies provide ongoing measurements so DD can resize ranges dynamically to relieve load hotspots and maintain balance [4][5].
```